_posts/2018-06-18-On-Obsessive-Statelessness.markdown

```diff
@@ -0,0 +1,201 @@
+can execute with the DSL provided in both [danger-js][] and [peril][ext]'s extensions. Peril's runtime is a
+reasonably normal nodejs environment, so it supports working with `node_modules` to get stuff done.
+
+This is where things get tricky, I first explored running code inside a tightened [virtual machine for node][vm2]
+but eventually found enough holes that it was definitely not going to work against malicious user-code in the same
+process as Peril. I lucked out to a potential answer to this when building out documentation for danger-js, which
+could [feed many birds with one bowl][birds]. I could separate out the execution context (think: the runtime DSL,
+the webhook JSON, and a bunch of installation specific config) as JSON and then pass that servers/processes then
+turn that back into a useful runtime again in a separate client which runs the Dangerfile (the name for a `js`/`ts`
+file running in the Peril JavaScript Environment. )
+
+This idea was so compelling that I first used it to create a version of Danger that runs
+[native to swift](https://github.com/danger/danger-swift) to figure out the kinks of what actually needs to be
+transmitted. For Peril, this meant I could explore having the evaluation of user-code inside a completely different
+server. I initially explored [using AWS Lambda][lambda] to run user-code, it's cheap, fast and mature. However, it's
+possible for lambda instances to communicate with each other, as each run is not [a fresh process][lambda-reuse].
+Making it not a secure platform for un-trusted code.
+
+Not deterred, I explored the world of docker hosting as a service - first exploring running my own cluster [on AWS
+ECS][ecs] and then settling [on Hyper][hyper] which offered sandboxed runs that booted in a few seconds. This is
+where my first real dive into obsessive statelessness comes in. The docker container, and the hyper environment
+contains no config by default. There is no Peril information available inside that runtime environment.
+
+The information about a Peril run comes exclusively from Peril, in my head, I call this dependency injection for the
+runtime environment. You can get a sense for what the full JSON looks like in this
+[fixtured file generated by tests](https://github.com/danger/peril/blob/master/source/github/events/handlers/_tests/fixtures/PerilRunnerEventBootStrapExample.json).
+It contains everything from (temporary) GitHub access tokens (only for your installation), to environment variables
+for your run and the webhook JSON. The runtime environment only knows that information for the duration of the
+process then all access tokens expires after 30 seconds of it starting regardless.
+
+## User Sessions are ephemeral
+
+With the runtime security figured out, and reasonably stable, I could start thinking about how people can control
+their installations inside peril. Taking ideas from Ashkan's [post on JWT's][jwts] I explored using JWTs to a few
```

Comment:
(Outdated question) Ah, up until this point, I wasn't aware of the existence of a user accessible Peril admin web service. (I'm super ignorant, haha.) It looks like below, there's a screenshot. That might be good to introduce up here.

(Outdated question) Just to be clear, is the basic idea that the Peril server gets identity info from an auth flow with GitHub, and then packages the minimum amount of that info into the JWT that is stored as a cookie in the admin's browser?

Ok, so looks like all the questions I had when reading are actually answered below. I think it might be worth revisiting the order of this client-side section to be something like:

1. Explain the existence and use of the admin tools.
2. Explain the reliance on GitHub as the source of authentication.
3. Explain the concept of using client-side storage for auth data.
4. Explain the use of JWT to represent and secure that data for storage in the unsafe space of the user's browser.